design/pipeline-caching.md

```diff
@@ -0,0 +1,157 @@
+# Pipeline Caching
+
+*** Status: Work in Progress ***
+
+We have observed the need for Azure Pipelines to implement a dependency caching mechanism that allows the outputs of steps within a pipeline to be optimized (or even skipped) if a suitable cached version of the outputs of those steps already exists. Caching can be effective provided the cost of determining a cache hit and acquiring the contents of the cache is cheaper than the cost of producing the output again from scratch.
+
+## Proposal
+
+This proposal contains a number of interconnected elements. We have attempted to layer them in this document from the most primitive building blocks up to simplified YAML syntax to make caching "just work" in the majority of cases. Our approach to building a caching mechanism for Azure Pipelines will be to get the fundamental building blocks right and prove we can increase build performance and then evolve the YAML syntax once we know exactly what that syntax will need to describe.
+
+### Restore and Save Cache Tasks
+
+The basic functions of Pipeline Cache will be implemented by two tasks. The ```RestoreCache``` task and the ```SaveCache``` task. Both tasks will specify a collection of files and environment variables which can be used as a key to lookup the cache and identify a hit or a miss. Each task will also take a collection of paths which will be fetched from or stored in the cache associated with that key.
+
+Consider the following Azure Pipelines pipeline defined in YAML:
+
+```yaml
+steps:
+- script: yarn
+- script: yarn build
+```
+
+In this example we are working on a React-based application which was bootstrapped using ```create-react-app```. The output of this command is a directory with about ~28,000 files including the ```node_modules``` directory. On a local developer machine the ```yarn``` command will trigger the installation of any missing packages. Typically this will execute quickly because all packages already exist on the developer machine. However in cloud-based CI environments where build machines are recycled after every build there is no persistent state which means that Yarn must download all packages from scratch.
+
+Applying caching to this pipeline would involve adding a restore cache step and save cache step before and after the ```-script: yarn``` step.
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    keys: |
+      package.json
+      yarn.lock
+    paths: |
+      node_modules
+- script: yarn
+- task: SaveCache@0
+  inputs:
+    keys: |
```

Comment:
Is there any way we can set this up so that "keys" and "paths" doesn't need to be duplicated?  Seems too easy for users to mess this up.

Reply:
After @chrisrpatterson comment about not including the paths in the restore step (meaning we need to store that info in the cache) I think the problem is less of an issue now. The restore step will just reference the key, the cache step will reference key and paths.

Reply:
If you are using key X and want to make it key X+Y, but you only update the store step, then you will continue to get cache hits on the restore step from when you had the earlier (inaccurate?) key.

Reply:
I think I made a mistake with the syntax here - it should have been ```key:``` not ```keys:``` I'm avoiding the multi-key scenario. I'm assuming that if we want to get fancy here in the future we'll get fancy along the lines of BuildXL and have a single weak fingerprint, and the service will return selectors to calculate a strong fingerprint. But that is out of scope for now.